Foundation/URLSession/http/HTTPURLProtocol.swift

```diff
@@ -60,6 +60,202 @@ internal class _HTTPURLProtocol: _NativeProtocol {
+                }
+                
+                return nil
+            }
+            
+            let parts = headerValue.components(separatedBy: ",").map { $0.trimmingCharacters(in: .whitespaces).lowercased(with: NSLocale.system) }
+            
+            for part in parts {
+                if part == "no-cache" {
+                    noCache = true
+                } else if part == "no-store" {
+                    noStore = true
+                } else if let maxAge = isWithArgument(part, named: "max-age", converter: { UInt($0) }) {
+                    self.maxAge = maxAge
+                } else if let sharedMaxAge = isWithArgument(part, named: "s-maxage", converter: { UInt($0) }) {
+                    self.sharedMaxAge = sharedMaxAge
+                }
+            }
+        }
+    }
+    
+    override func canCache(_ response: CachedURLResponse) -> Bool {
+        guard let httpRequest = task?.currentRequest else { return false }
+        guard let httpResponse = response.response as? HTTPURLResponse else { return false }
+        
+        let now = Date()
+        
+        // Figure out the date we should start counting expiration from.
+        let expirationStart: Date
+        
+        if let dateString = httpResponse.allHeaderFields["Date"] as? String,
+           let date = _HTTPURLProtocol.dateFormatter.date(from: dateString) {
+            expirationStart = min(date, response.date) // Do not accept a date in the future of the point where we stored it, or of now if we haven't stored it yet. That is: a Date header can only make a response expire _faster_ than if it was issued now, and can't be used to prolong its age.
+        } else {
+            expirationStart = response.date
+        }
+        
+        // We opt not to cache any requests or responses that contain authorization headers.
+        if httpResponse.allHeaderFields["WWW-Authenticate"] != nil ||
+           httpResponse.allHeaderFields["Proxy-Authenticate"] != nil ||
+           httpRequest.allHTTPHeaderFields?["Authorization"] != nil ||
+           httpRequest.allHTTPHeaderFields?["Proxy-Authorization"] != nil {
+            return false
+        }
+        
+        // HTTP Methods: https://tools.ietf.org/html/rfc7231#section-4.2.3
+        switch httpRequest.httpMethod {
+        case "GET":
+            break
+        case "HEAD":
+            if response.data.isEmpty {
+                break
+            } else {
+                return false
+            }
+        default:
+            return false
+        }
+        
+        // Cache-Control: https://tools.ietf.org/html/rfc7234#section-5.2
+        var hasCacheControl = false
+        var hasMaxAge = false
+        if let cacheControl = httpResponse.allHeaderFields["Cache-Control"] as? String {
+            let directives = CacheControlDirectives(headerValue: cacheControl)
+            
+            if directives.noCache || directives.noStore {
+                return false
```

Comment:
My understanding of `no-cache` is that the caches should revalidate with the origin server, but the cached content can still be used if the server answers “Not modified”.